Generate build status-dependent icons for the editor

tehKaiN opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/541

Also auto-generates the icons for console executables under Windows. They are not pixel-exact to the previous ones, since they're composited dynamically, but they should be close enough.

Sample preview using some kind of web-based .ico previewer - might be a bit distorted.

![main_full](https://github.com/user-attachments/assets/871c480d-9af1-42ef-86bc-71e32952aa74)
![main_full](https://github.com/user-attachments/assets/e9185d65-a890-446d-b73e-08be3ccd1c92)

Caveats:
- Icon components are rendered to PNG first, then composited, because doing everything in SVG and then converting to PNG yielded bad results - some lines were thinner than others, also getting color-mixed with the neighbors.
- I'm mainly a programmer, so SVGs present in the PR are not perfect - though I assume they are good enough placeholders. I've concocted them in GodSVG in conjunction with some kind of web-based SVG text path generator.
- No linux icons because I couldn't get any response on how to tackle it in the proposal issue. :(

Also it made a bit of make_icons.sh obsolete, so perhaps it could supersede it further? I don't know how it's called though - make_icons.py is intended to generate files which are to be commited, whereas it looks like make_icons.sh is used for packaging?

tehKaiN committed:
Generate build status-dependent icons for android/macos/windows versions of the editor

bruvzg commented on platform/macos/SCsub:
```diff
@@ -129,6 +131,12 @@ files = [
 
 prog = env.add_program("#bin/godot", files)
 
+env.Command(
+    "#misc/dist/macos_tools.app/Contents/Resources/Godot.icns",
+    "#platform/macos/icons/Godot_" + env.version_info["status"] + ".icns",
+    Copy("$TARGET", "$SOURCE"),
+)
```
Probably should be copied in the `generate_bundle` instead (directly to target bundle), icon is not used outside of bundle.

bruvzg reviewed (commented):
There's also embedded PNG icon which is used for window title / taskbar / task switcher:

https://github.com/godotengine/godot/blob/a3080477ac0421aef24ca0916c40559abbf4846b/main/SCsub#L35-L40

tehKaiN commented:
> The script for generating PNGs is pretty much all that can be done, Linux icon is not part of the executable or official distribution, so it's up to distro specific package maintainers how to use it.

That should be easily doable, but I need to know what kind of PNG sizes are needed.

bruvzg commented:
> That should be easily doable, but I need to know what kind of PNG sizes are needed.

No sure if there's a standard set of sizes. The following seems to be used `16x16, 22x22, 24x24, 32x32, 48x48, 64x64, 96x96, 128x128, 192x192, 256x256`, but not every icon is using all of them. Some apps also use single SVG instead.

tehKaiN commented:
I've added a bunch of linuxbsd icons. Also generates the SVG variant. The problem is, generated icons don't pass through svgo, but since the pre-commit hook generates the optimized variant, and the script doesn't need to run that often, perhaps that's acceptable? If not then there's a problem - I see no better SVG libary than cairo and I'm not quite sure if text-based svg editing won't trigger the svgo anyway.

tehKaiN commented:
What's the status of this PR? I know it's probably of very low priority, but perhaps there are some concerns that I could investigate in the meantime.

bruvzg commented on platform/macos/SCsub:
```diff
@@ -38,6 +40,10 @@ def generate_bundle(target, source, env):
         if "mono" in env.module_version_string:
             shutil.copytree(Dir("#bin/GodotSharp").abspath, app_dir + "/Contents/Resources/GodotSharp")
+        shutil.copy(
+            Dir("#platform/macos/icons/Godot_" + env.version_info["status"] + ".icns").abspath,
```
```suggestion
            File("#platform/macos/icons/Godot_" + env.version_info["status"] + ".icns").abspath,
```
It's file, not directory.

bruvzg reviewed (approved):
macOS and Windows icons look good to me.

m4gr3d reviewed (approved):
The Android section looks good.

syntaxerror247 commented:
@tehKaiN Can you provide Android icon_monochrome for different sizes too? 
This is the svg file, you can resize it: https://godotengine.org/assets/press/icon_monochrome_dark.svg

tehKaiN commented:
I didn't know that they are required/nice to have since the original icon set didn't had those. Fixed!
Regarding using SVG, if the quality of current icons isn't good enough then I'll investigate going the SVG route, but that needs extra work to make the dev/beta/rc label svgs monochrome too.

syntaxerror247 commented:
> Regarding using SVG, if the quality of current icons isn't good enough then I'll investigate going the SVG route, but that needs extra work to make the dev/beta/rc label svgs monochrome too.

I had suggested using the SVG as a base image, resizing it to your desired dimensions, and then exporting it as a PNG file.

tehKaiN commented:
A friendly reminder that another month passed by. I know it's super low priority issue, but I wonder - is there some kind of review queue where I could check how things progress? Or should I just wait for the reviews, or perhaps keep pinging here and there from time to time?

tehKaiN commented:
Updated the PR for post-4.4 changes.
One thing I see is that 4.4 added android icon_monochrome.png for all the DPI variants. I've removed them since they are already generated in my PR.

One thing that's curious for me is that 4.4 changed the icon_foreground and icon_monochrome proportions in mdpi: previously, and in my PR, it was 66px square icon in 108px transparent square background as per [Android guidelines](https://developer.android.com/develop/ui/views/launch/icon_design_adaptive#design-adaptive-icons), but now the icon is something around 48px square. It would be nice to have a second opinion on how it should really be.

syntaxerror247 commented:
@tehKaiN icon_foreground haven't been updated for last 6 years.

tehKaiN commented:
Sorry, it was late yesterday. Anyway, there is discrepancy between proportions of those icons between current version and my version, and I'm not sure whether I should strictly follow what's in the repo currently, or keep it as I've changed it since android docs seem to confirm that latter is better. I'm not an android developer though, so I might got that wrong.
